src/.prototype/musicutils/KeySignature.ts

```diff
@@ -0,0 +1,287 @@
+            if (!this.preferSharps(this.key, this.mode) && this.key.indexOf('#') !== -1) {
+                scale.forEach((note: string, i: number) => {
+                    if (note.indexOf('b') != -1) {
+                        if (note in MusicUtils.EQUIVALENT_SHARPS) {
+                            scale[i] = MusicUtils.EQUIVALENT_SHARPS[note];
+                        }
+                    }
+                });
+            }
+
+            // For Latin scales, we cannot skip notes.
+            if (scale.length == 8) {
+                for (let i = 0; i < scale.length - 1; i++) {
+                    let idx1: number = MusicUtils.PITCH_LETTERS.indexOf(scale[i][0]);
+                    let idx2: number = MusicUtils.PITCH_LETTERS.indexOf(scale[i + 1][0]);
+                    if (idx2 < idx1) {
+                        idx2 += 7;
+                    }
+                    if (idx2 - idx1 > 1) {
+                        if (scale[i + 1] in MusicUtils.CONVERT_DOWN) {
+                            scale[i + 1] = MusicUtils.CONVERT_DOWN[scale[i + 1]];
+                        }
+                    }
+                }
+            }
+            // And ensure there are no repeated letter names.
+            for (let i = 0; i < scale.length - 1; i++) {
+                let new_current_note: string, new_next_note: string;
+                if (i === 0 && scale[i][0] === scale[i + 1][0]) {
+                    if (scale[i + 1] in MusicUtils.CONVERT_UP) {
+                        const new_next_note: string = MusicUtils.CONVERT_UP[scale[i + 1]];
+                        scale[i + 1] = new_next_note;
+                    }
+                } else if (scale[i][0] === scale[i + 1][0]) {
+                    if (scale[i] in MusicUtils.CONVERT_DOWN) {
+                        new_current_note = MusicUtils.CONVERT_DOWN[scale[i]];
+                    } else {
+                        console.log(scale[i]);
+                    }
+                    // If changing the current note makes it the same
+                    // as the previous note, then we need to change the
+                    // next note instead.
+                    if (new_current_note[0] != scale[i - 1][0]) {
+                        scale[i] = new_current_note;
+                    } else {
+                        if (scale[i + 1] in MusicUtils.CONVERT_UP) {
+                            new_next_note = MusicUtils.CONVERT_UP[scale[i + 1]];
+                            scale[i + 1] = new_next_note;
+                        }
+                    }
+                }
+            }
+        } else {
+            // Convert to preferred accidental.
+            if (this.key.indexOf('#') !== -1) {
+                scale.forEach((note: string, i: number) => {
+                    if (note.indexOf('b') !== 0) {
+                        if (note in MusicUtils.EQUIVALENT_SHARPS) {
+                            scale[i] = MusicUtils.EQUIVALENT_SHARPS[note];
```

Comment:
What to you think about encapsulating exports from modules like util modules that export a lot of things into a single entity? I think it makes the code a little more readable as it immediately clear where something comes from.

Reply:
Surely there's better organization possible like you suggest. I'll take a dab at the classes and get back later today.

Reply:
Integrating all the exports as a JSON at the end of the file for long files seems like a better approach to me. Kind of gives the interface-like feel.